processor/memorylimiter/memorylimiter.go

```
@@ -0,0 +1,208 @@
+	if checkInterval <= 0 {
+		return nil, errCheckIntervalOutOfRange
+	}
+	if memAllocLimit == 0 {
+		return nil, errMemAllocLimitOutOfRange
+	}
+	if memSpikeLimit >= memAllocLimit {
+		return nil, errMemSpikeLimitOutOfRange
+	}
+
+	ml := &memoryLimiter{
+		traceConsumer:   traceConsumer,
+		metricsConsumer: metricsConsumer,
+		memAllocLimit:   memAllocLimit,
+		memSpikeLimit:   memSpikeLimit,
+		memCheckWait:    checkInterval,
+		ballastSize:     ballastSize,
+		ticker:          time.NewTicker(checkInterval),
+		readMemStatsFn:  runtime.ReadMemStats,
+		statsTags:       statsTagsForBatch(name),
+	}
+
+	initMetrics()
+
+	ml.startMonitoring()
+
+	return ml, nil
+}
+
+func (ml *memoryLimiter) ConsumeTraceData(
+	ctx context.Context,
+	td consumerdata.TraceData,
+) error {
+
+	if ml.forcingDrop() {
+		numSpans := len(td.Spans)
+		stats.RecordWithTags(
+			context.Background(),
+			ml.statsTags,
+			StatDroppedSpanCount.M(int64(numSpans)))
+
+		return errForcedDrop
+	}
+	return ml.traceConsumer.ConsumeTraceData(ctx, td)
+}
+
+func (ml *memoryLimiter) ConsumeMetricsData(
+	ctx context.Context,
+	md consumerdata.MetricsData,
+) error {
+
+	if ml.forcingDrop() {
+		numMetrics := len(md.Metrics)
+		stats.RecordWithTags(
+			context.Background(),
+			ml.statsTags,
+			StatDroppedMetricCount.M(int64(numMetrics)))
+
+		return errForcedDrop
+	}
+	return ml.metricsConsumer.ConsumeMetricsData(ctx, md)
+}
+
+func (ml *memoryLimiter) GetCapabilities() processor.Capabilities {
+	return processor.Capabilities{MutatesConsumedData: false}
+}
+
+func (ml *memoryLimiter) Start(host component.Host) error {
+	return nil
+}
+
+func (ml *memoryLimiter) Shutdown() error {
+	ml.ticker.Stop()
+	return nil
+}
+
+func (ml *memoryLimiter) readMemStats(ms *runtime.MemStats) {
+	ml.readMemStatsFn(ms)
+	// If proper configured ms.Alloc should be at least ml.ballastSize but since
+	// a misconfiguration is possible check for that here.
+	if ms.Alloc >= ml.ballastSize {
+		ms.Alloc -= ml.ballastSize
+	}
```

Should the else be logged as a warning?

If subtracting `ml.ballastSize` from m1.Alloc is silently ignored it could cause unintended behavior of this processor.  It would also help the operator notice if their were drift between the ballast size configuration for this processor and the command line `mem-ballast-size-mib` option.

---

Good point. Added a one-time log warning on `else`.